### ISSUE TITLE
01-intro: Inaccuracy in Demonstration of "Variables as Sticky Notes"

### ISSUE THREAD

**Issue by bielsnohr** [swcarpentry/python-novice-inflammation#847]
First off, let me start by saying I am a big fan of the Carpentries ever since I attended one of the workshops at the beginning of my PhD. Truly shaped how I approached software development. Many years later, I am finally in a position to start contributing to the lessons, and I am hoping to become an instructor soon. Now to the technical stuff.

A callout in the first episode (01-intro) seeks to describe ["Variables as Sticky Notes"](https://github.com/swcarpentry/python-novice-inflammation/blame/gh-pages/_episodes/01-intro.md#L137). I think this is a good analogy for Python variables, but the subsequent example used to support this actually demonstrates an entirely different characteristic of Python.

The call out finishes by saying:
> Since `weight_lb` doesn't "remember" where its value comes from,
> it is not updated when we change `weight_kg`.

Variables "not remembering" where they come from and therefore not being updated when other (seemingly related) variables are changed has nothing to do with variables behaving like labels and everything to do with eager/lazy execution. 

Python uses eager execution, so the variable assignment
```python
weight_lb = 2.2 * weight_kg
```
is done immediately, meaning the value that `weight_kg` points to is retrieved and used to calculate the value that `weight_lb` is stuck onto (to extend the sticky note analogy). `weight_lb` doesn't store the relationship to `weight_kg` not because they are both labels/sticky notes, but because the execution of the assignment happens immediately.

I think a possibly better example is something like the following
```python
>>> my_weight = 60
>>> friend_weight = my_weight
>>> my_weight is friend_weight
True

>>> my_weight == friend_weight
True

>>> hex(id(my_weight))
'0x55ce337fba80'

>>> hex(id(friend_weight))
'0x55ce337fba80'

>>> my_weight = 72
>>> friend_weight
60

>>> hex(id(my_weight))
'0x55ce337fbc00'

>>> hex(id(friend_weight))
'0x55ce337fba80'
```

This could then be contrasted to what the result would be in C where variables are like buckets rather than labels. A good explanation I come back to is this one: <https://realpython.com/pointers-in-python/#variables-in-c>. The key difference is the memory location comparisons in C would not match like they do in Python at the beginning of this example. Moreover, the memory location would not change upon `my_weight = 72` being executed.

However, I appreciate that this get quite involved and technical, more than might be reasonable for an introductory course. Unfortunately, Python variables are quite weird, and it is really difficult to find short examples that clearly display the label/sticky note nature of them.

**Comment by bielsnohr** [comment]
Looking at the other recent issues submitted, it seems like mine is the only one that hasn't received a comment. Is there anything I can do to help move this issue along? Shall I submit a pull request instead?

**Comment by maxim-belkin** [comment]
Hi Matthew,

Apologies for dropping the ball here and thanks for submitting this issue! ❤️  Please see my response below.

Maxim

---- 

> Variables "not remembering" where they come from and therefore not being updated when other (seemingly related) variables are changed has nothing to do with variables behaving like labels and everything to do with eager/lazy execution.

We use "variables as sticky notes" analogy to give learners unfamiliar with programming basic idea about variables. The episode doesn't say that all aspects of their behavior can be explained using this analogy. However, I think that the part  you're talking about specifically can be explained using "sticky notes" analogy: we create a new "sticky note" whenever there is an equals sign. Here we don't talk about when we do the calculation implying that it is done immediately -- perhaps, we should do that. I also think that we need to update our image to use the sticky note images we all know -- colored squares -- with values written in big letters in the center and variable names written in the top left (or right)-hand corner -- because in Python variables can't exist without a value, which is implied by our current image.

Of course, we can't go into such a low-level stuff as memory addresses, pointers, comparison to C, etc. -- this is not the goal of this section, which is to explain variables in normal human language.

**Comment by nihiliad** [comment]
I agree with Matthew that this example could be easier to understand, but for reasons other than whatever its technical accuracy may be. In addition to "Variables as Sticky Notes", which I think is a great metaphor, there are a few other things going on with this example that could distract people from that main point:

- `weight_kg` and `weight_lb` refer to an equivalent weight, but their units, and therefore their values, are different
- Assigning an equivalent weight in pounds requires a conversion from weight in kilograms
- The above conversion makes the assignment more complicated, involving not just `weight_kg` but a mathematical expression

That's a lot to keep in mind in addition to the concept of variables as sticky notes. If we used something other than weight, we could make it simpler. Maybe something like this?

Kemal is 25 years old: 
`age_kemal = 25`
(Image with an "age_kemal" sticky note on the value 25.)

Sarah is the same age as Kemal:
`age_sarah = age_kemal`
(Image with two sticky notes, "age_kemal" and "age_sarah", both on the value 25.)

Kemal has a birthday:
`age_kemal = 26`
(Two images, one with sticky note "age_kemal" on the value 26, the other with "age_sarah" on 25.)

We could still keep the more complicated weight example, but maybe move it into an exercise.

**Comment by bielsnohr** [comment]
> Apologies for dropping the ball here and thanks for submitting this issue! :heart: 

@maxim-belkin not a problem at all! I was mostly worried I had done something wrong. I also appreciate that most people here are probably doing this in there free/spare time, so I had no expectation of service :wink: 

> we create a new "sticky note" whenever there is an equals sign

Even in the context of the fuzzy analogy, I don't think this is true. For the initial binding of a name/variable, then sure the "sticky note" is created at this point; however for subsequent bindings of the name, it is the same "sticky note", just stuck to different values. Whether a "sticky note" or value is created with the equals operator depends on what is in memory. Example

```python
>>> x = 1  # "sticky note" x is created and put on value 1
>>> x = 2  # "sticky note" x is put on value 2, no need to create the label this time
>>> x = x + 1  # "sticky note" x is put on value 2 + 1 = 3, again no need to create the sticky note
```

> Here we don't talk about when we do the calculation implying that it is done immediately -- perhaps, we should do that.

Yes, I think it would be beneficial to mention when binding and calculations happen, because as I said in my initial message, this is the reason that Python names (and variables in most languages generally) don't "remember" where they have come from. C variables also don't remember where they come from, but it has a completely different memory model from Python. The point being that "remembering" has to do with delayed execution and not the particularly schema for variables.

> I also think that we need to update our image to use the sticky note images we all know -- colored squares -- with values written in big letters in the center and variable names written in the top left (or right)-hand corner -- because in Python variables can't exist without a value, which is implied by our current image.

I also have to disagree here. The "sticky note" (i.e. variable/name) is distinct from the value (i.e. PyObject) to which is it stuck. Having the value on the sticky note itself will conflate this, and in my view, confuse the nature of Python names/variables that we are trying to explain. It sounds like we have a fundamentally different perception of what the sticky note analogy means in practice. Based on what actually goes on under the hood in Python, I believe it is more accurate to state that the variable/name (say `x`) constitutes the sticky note in its entirety, and then this sticky note is moved between various values as calculations and bindings are performed. This is closer to the current illustrations in the lesson. I could write a mock of this up here if you think it would be valuable?

> Of course, we can't go into such a low-level stuff as memory addresses, pointers, comparison to C, etc. -- this is not the goal of this section, which is to explain variables in normal human language.

Certainly, I don't think it would be good get into the nuts and bolts, but at the same time, I think it is pretty important to get some idea of the Python object model. This might be done by comparing it at a high level to the analogy for C variables which are like buckets rather than labels/sticky notes. The important implication is that Python names can point to the same object in memory while C variables (not pointers) will each have their own distinct memory location. This doesn't really make a difference for immutable types like integers, but it becomes very important for mutable types like lists/arrays.

**Comment by bielsnohr** [comment]
> Sarah is the same age as Kemal:
`age_sarah = age_kemal`
(Image with two sticky notes, "age_kemal" and "age_sarah", both on the value 25.)

@nihiliad This is exactly the sort of behaviour of Python names/variables that I think we should be trying to emphasise (i.e. that they can point to the same value in memory). As your example shows, this would actually be simpler than what is presently in the call out.

**Comment by nihiliad** [comment]
Thanks, @bielsnohr! I also like the way the lesson uses the "sticky note" metaphor, and agree with what you said here:

> The "sticky note" (i.e. variable/name) is distinct from the value (i.e. PyObject) to which is it stuck. Having the value on the sticky note itself will conflate this, and in my view, confuse the nature of Python names/variables that we are trying to explain. It sounds like we have a fundamentally different perception of what the sticky note analogy means in practice. Based on what actually goes on under the hood in Python, I believe it is more accurate to state that the variable/name (say x) constitutes the sticky note in its entirety, and then this sticky note is moved between various values as calculations and bindings are performed. This is closer to the current illustrations in the lesson.

**Comment by ldko** [comment]
Interesting discussion about this sticky note issue @bielsnohr, @nihiliad , @maxim-belkin . I can see why attributing "This means that assigning a value to one variable does not change values of other variables" to the sticky note concept itself could be misleading. As Maxim suggests, we are using the analogy to give a general idea of variables to people who may have no programming experience and are in their first few minutes of the lesson. I don't think we need to go into memory locations and how variables in Python work differently from C. While this might be a good way to explain things to someone taking a  full programming course or working to become a professional programmer, we have a handful of hours to show our learners how to use code to automate producing visualizations of their data. While it may seem we are skipping over important details of how Python is implemented and how it compares to other languages, it out of scope of this lesson to go into many of the details and explanations we would cover in a true Python programming course.

I wonder if we could fix the issue by tweaking the wording in the callout to suggest we are using the sticky note concept to track variable values in the example rather than to imply that sticky notes are the reason the variables behave as they do.

After this:

>A variable is analogous to a sticky note with a name written on it: assigning a value to a variable is like putting that sticky note on a particular value.

Then perhaps we could change the wording of the next text to something like:

> Using this analogy, let's look at how assigning a value to one variable does *not* change values of other variables.  For example, let’s store the subject’s weight in pounds in its own variable:

Then maybe after the code explain:

> First, the calculation of 2.2 * 65.0 (the value of `weight_kg`) is carried out,
then the resulting value is assigned the variable name.

I kind of like the idea of the age example and seeing multiple variable names stuck on one value as @nihiliad brought up. However, it seems if we did highlight this feature, when we later talk about mutable vs. immutable in the lists episode, we might want to bring back the sticky note analogy to show how it doesn't work the same way there. I am not sure if this would potentially overall cause more confusion overall for those new to programming though.

**Comment by bielsnohr** [comment]
Hi @ldko and thank you for the response. Apologies it has taken me so long to return to this.

> While this might be a good way to explain things to someone taking a full programming course or working to become a professional programmer, we have a handful of hours to show our learners how to use code to automate producing visualizations of their data. While it may seem we are skipping over important details of how Python is implemented and how it compares to other languages, it out of scope of this lesson to go into many of the details and explanations we would cover in a true Python programming course.

Yes, given this context, you are certainly right that it is probably out of scope to compare Python variables to those in other languages. If we do ever decide this might be valuable, then I think it could be kept fairly high-level using analogies (i.e. Python names are like sticky-notes or labels that you stick to values, whereas C variables are like buckets that you put values in). Personally, this was a very important lesson along my way to proficiency in both Python and C.

>> A variable is analogous to a sticky note with a name written on it: assigning a value to a variable is like putting that sticky note on a particular value.
>
>Then perhaps we could change the wording of the next text to something like:
>> Using this analogy, let's look at how assigning a value to one variable does not change values of other variables. For example, let’s store the subject’s weight in pounds in its own variable:
>
> Then maybe after the code explain:
>> First, the calculation of 2.2 * 65.0 (the value of weight_kg) is carried out, then the resulting value is assigned the variable name.

This sounds like a sensible modification to eliminate misleading learners. 

> I kind of like the idea of the age example and seeing multiple variable names stuck on one value as @nihiliad brought up. However, it seems if we did highlight this feature, when we later talk about mutable vs. immutable in the lists episode, we might want to bring back the sticky note analogy to show how it doesn't work the same way there. I am not sure if this would potentially overall cause more confusion overall for those new to programming though.

Actually, @nihiliad's example will behave exactly the same regardless of whether the PyObjects they refer to are mutable or immutable. Take the analogue with lists:

```python
ages_class_1 = [10, 11, 10, 12]
ages_class_2 = ages_class_1
# someone has a birthday in class 1
ages_class_1 = [10, 12, 10, 12]
```

As in the previous case with integers, `ages_class_1` and `ages_class_2` will point to different PyObject lists. What I imagine you are concerned about is something like this:


```python
ages_class_1 = [10, 11, 10, 12]
ages_class_2 = ages_class_1
# someone has a birthday in class 1
ages_class_1[1] += 1
```

This will behave differently than:

```python
age_kemal = 10
age_sarah = age_kemal
# Kemal has a birthday
age_kemal += 1
```

In the former, `ages_class_1` and `ages_class_2` will point to the same list, whereas `age_kemal` and `age_sarah` will not point to the same integer value. For the immutable vs mutable callout in lesson 5, it might actually be valuable to bring back the sticky-note analogy. The operations on the individual elements of the list does not alter the fact that the name will still point to that same list. Indeed, this is just the point that is made in that section anyway:

> Be careful when modifying data in-place. If two variables refer to the same list, and you modify the list value, it will change for both variables!

In conclusion (phew, sorry about the length) I would recommend we move to using the age example because it is cleaner and simpler.

**Comment by bielsnohr** [comment]
After quite a bit of delay, I have finally gotten around to submitting a pull request related to this. As a stop-gap, I have put the suggestions by @ldko into the lesson. But it is still my feeling that an example like the one discussed above should eventually replace the current one. This will probably involve some changes to the list episode such that it also refers back to this. If I have further time, I will consider drafting that up. Thanks to all who have contributed on this issue.
